### PR TITLE
use async azure-iot-sdk

### PR DESCRIPTION
**Pull request by JanZachmann**
- replaced channel/thread based event dispatching by async dispatching based on azure-iot-sdk 0.10.0
 - introduced mockall crate for unit tests
 - introduced cargo feature "mock"
   - usefull for host testing
   - automatically used by cargo test

### CODE REVIEW

File: src/twin/mod_test.rs [hunk 1]
```diff
@@ -1,62 +1,82 @@
+/*
+    call tests always single threaded since otherwise setting special test env vars clashes:
+
+    'cargo test -- --test-threads=1'
```
Comment (empwilli) [inline src/twin/mod_test.rs:4]: fwii: we could think about enforcing this option in the build.rs: 
https://github.com/rust-lang/cargo/issues/8430#issuecomment-1547668554

File: src/twin/mod_test.rs [hunk 2]
```diff
@@ -1,62 +1,82 @@
...
+                tx_connection_status: Option<mpsc::Sender<AuthenticationStatus>>,
+                tx_twin_desired: Option<mpsc::Sender<(TwinUpdateState, serde_json::Value)>>,
+                tx_direct_method: Option<DirectMethodSender>,
+                tx_incoming_message: Option<IncomingMessageObserver>,
+            ) -> Result<Box<Self>>;
+            fn twin_async(&mut self) -> Result<()>;
+            async fn send_d2c_message(&mut self, mut message: IotMessage) -> Result<()>;
+            async fn twin_report(&mut self, reported: serde_json::Value) -> Result<()>;
+        }
+
```
Comment (empwilli) [inline src/twin/mod_test.rs:51]: nit: drop newline

File: Cargo.toml [hunk 3]
```diff
@@ -35,17 +36,20 @@ serde_with = "2.2"
 strum = "0.24"
 strum_macros = "0.24"
 systemd-zbus = "0.1"
-# ToDo: set to 0.3 once we support rust 1.63
-time = { version = "=0.3.19", features = ["formatting"] }
+time = { version = "=0.3.23", features = ["formatting"] }
```
Comment (empwilli) [inline Cargo.toml:39]: Do we still need explicitly this version? Why so?

**Review by empwilli — CHANGES_REQUESTED** [review]
Some minor nits.

File: Cargo.toml (hunk 3 above)
Comment (JanZachmann) [reply Cargo.toml:39]: we need was is supported by 1.65 
afaik 0.3.24 requires 1.67

File: src/twin/mod.rs [hunk 4]
```diff
@@ -168,92 +74,89 @@ trait Feature {
...
                     TypeId::of::<WifiCommissioning>(),
-                    Box::new(WifiCommissioning::default()) as Box<dyn Feature + Send>,
+                    Box::<WifiCommissioning>::default() as Box<dyn Feature>,
                 ),
             ]),
         }
     }
 
-    pub fn init_features(&mut self) -> Result<()> {
+    pub async fn init_features(&mut self) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply src/twin/mod.rs:129]: rename to just `init`

**Review by empwilli — APPROVED** [review]

**Pull request merged**